File: _attic/bane/core/marshal/enums.go
package marshal

import (
	"errors"
	"fmt"
	"reflect"

	"github.com/wrmsr/bane/core/maps"
	rfl "github.com/wrmsr/bane/core/reflect"
)

///

type EnumMarshaler[T comparable] struct {
	m map[T]Value
}

func NewEnumMarshaler[T comparable](m map[T]string) EnumMarshaler[T] {
	return EnumMarshaler[T]{m: maps.MapValues(func(s string) Value { return MakeString(s) }, m)}
}

var _ Marshaler = EnumMarshaler[int]{}

func (m EnumMarshaler[T]) Marshal(ctx *MarshalContext, rv reflect.Value) (Value, error) {
	iv := rv.Interface()
	it, ok := iv.(T)
	if !ok {
		return nil, unhandledTypeOf(rv.Type())
	}
	mv, ok := m.m[it]
	if !ok {
		return nil, errors.New("unknown enum value")
	}
	return mv, nil
}

//

type EnumUnmarshaler struct {
	m map[string]reflect.Value
}

func NewEnumUnmarshaler(m map[string]any) EnumUnmarshaler {
	return EnumUnmarshaler{m: maps.MapValues(reflect.ValueOf, m)}
}

var _ Unmarshaler = EnumUnmarshaler{}

func (u EnumUnmarshaler) Unmarshal(ctx *UnmarshalContext, mv Value) (reflect.Value, error) {
	switch mv := mv.(type) {

	case String:
		s := mv.v
		v, ok := u.m[s]
		if !ok {
			return rfl.Invalid(), fmt.Errorf("unknown enum value: %s", s)
		}
		return v, nil

	}
	return rfl.Invalid(), unhandledTypeOf(reflect.TypeOf(mv))
}

///

func SetEnumTypes[T comparable](m map[T]string) SetType {
	i := make(map[string]any, len(m))
	for k, v := range m {
		i[v] = k
	}
	return SetType{
		Marshaler:   NewEnumMarshaler[T](m),
		Unmarshaler: NewEnumUnmarshaler(i),
	}
}

func SetStringerEnumTypes[T comparable](vs ...T) SetType {
	m := make(map[T]string, len(vs))
	for _, v := range vs {
		var o any
		o = v
		s := o.(fmt.Stringer).String()
		m[v] = s
	}
	return SetEnumTypes[T](m)
}

File: omlish/marshal/enums.py
import dataclasses as dc
import enum
import typing as ta

from .. import check
from .. import matchfns as mfs
from .. import reflect as rfl
from .base import MarshalContext
from .base import Marshaler
from .base import MarshalerFactory
from .base import UnmarshalContext
from .base import Unmarshaler
from .base import UnmarshalerFactory
from .values import Value


@dc.dataclass(frozen=True)
class EnumMarshaler(Marshaler):
    ty: type[enum.Enum]

    def marshal(self, ctx: MarshalContext, o: ta.Any) -> Value:
        return o.name


class EnumMarshalerFactory(MarshalerFactory):
    def guard(self, ctx: MarshalContext, rty: rfl.Type) -> bool:
        return isinstance(rty, type) and issubclass(rty, enum.Enum)

    def fn(self, ctx: MarshalContext, rty: rfl.Type) -> Marshaler:
        if not isinstance(rty, type) and issubclass(rty, enum.Enum):
            raise mfs.MatchGuardError(ctx, rty)
        return EnumMarshaler(rty)


@dc.dataclass(frozen=True)
class EnumUnmarshaler(Unmarshaler):
    ty: type[enum.Enum]

    def unmarshal(self, ctx: UnmarshalContext, v: Value) -> ta.Any:
        return self.ty[check.isinstance(v, str)]


class EnumUnmarshalerFactory(UnmarshalerFactory):
    def guard(self, ctx: UnmarshalContext, rty: rfl.Type) -> bool:
        return isinstance(rty, type) and issubclass(rty, enum.Enum)

    def __call__(self, ctx: UnmarshalContext, rty: rfl.Type) -> Unmarshaler | None:

        if isinstance(rty, type) and issubclass(rty, enum.Enum):
            return EnumUnmarshaler(rty)
        return None
